src/fs/bundle.rs

```diff
@@ -0,0 +1,235 @@
+            BundleData::Zip { zip, bundle_path } => {
+                let archive = Arc::new(Mutex::new(zip));
+
+                let mut archive_guard = archive.lock().unwrap();
+
+                let mut builder = FsNodeBuilder::new();
+                for i in 0..archive_guard.len() {
+                    let file = archive_guard.by_index(i).unwrap(); // TODO: report IO error?
+                    let name = file.name();
+                    if let Some(path) = name.strip_prefix(&bundle_path) {
+                        let path = GuestPath::new(path);
+                        if file.is_dir() {
+                            builder.add_directory(&path);
+                        } else {
+                            builder.add_file(
+                                &path,
+                                FsNode::bundle_zip_file(IpaFileRef {
+                                    archive: archive.clone(),
+                                    index: i,
+                                }),
+                            );
+                        }
+                    }
+                }
+                builder.build()
+            }
+        }
+    }
+
+    pub fn read_plist(&mut self) -> Result<Vec<u8>, std::io::Error> {
+        match self {
+            BundleData::HostDirectory(path) => std::fs::read(path.join("Info.plist")),
+            BundleData::Zip { zip, bundle_path } => {
+                let mut file = zip.by_name(&format!("{}/Info.plist", bundle_path))?;
+                let mut buf = Vec::new();
+                file.read_to_end(&mut buf)?;
+                Ok(buf)
+            }
+        }
+    }
+}
+
+/// Represents a file inside an IPA bundle that can be opened.
+#[derive(Debug)]
+pub struct IpaFileRef {
+    archive: Arc<Mutex<ZipArchive<std::fs::File>>>,
+    index: usize,
+}
+
+impl IpaFileRef {
+    pub fn open(&self) -> IpaFile {
+        let mut archive = self.archive.lock().unwrap();
+        let mut file = match archive.by_index(self.index) {
+            Ok(file) => file,
+            Err(ZipError::Io(e)) => {
+                // this is a runtime error, which we __probably__ should not bubble up to the guest
+                panic!("IO error while opening file from IPA bundle: {e}")
+            }
+            // anything other than IO error is a bug in the code, we should always have a valid index
+            Err(e) => panic!("BUG: could not open file from IPA bundle: {e}"),
+        };
+        let mut buf = Vec::new();
+        file.read_to_end(&mut buf).unwrap();
+        IpaFile {
+            file: std::io::Cursor::new(buf),
+        }
+    }
+}
+
+/// Represents an opened file in an IPA bundle.
+pub struct IpaFile {
+    // we need to use a cursor because zip::read::ZipFile doesn't implement Seek
+    // and, generally, seeking in compressed files is hard to achieve
+    // the simplest way to do it is to read the whole file into memory
+    // the target apps should be small enough to fit in memory, right?
```

Comment:
Hmm, this might end up a problem with large files (app bundles can contain video files and stuff), but I assume they won't be kept open for too long so it sounds fine for now.

Reply:
Yeah, I figures this too. Just something to be aware of